electrum/onion_message.py

```diff
@@ -0,0 +1,742 @@
+    shared_secret = get_ecdh(privkey, blinding)
+    b_hmac = get_bolt04_onion_key(b'blinded_node_id', shared_secret)
+    b_hmac_int = int.from_bytes(b_hmac, byteorder="big")
+
+    our_privkey_int = int.from_bytes(privkey, byteorder="big")
+    our_privkey_int = our_privkey_int * b_hmac_int % ecc.CURVE_ORDER
+    our_privkey = our_privkey_int.to_bytes(32, byteorder="big")
+
+    return our_privkey
+
+
+def is_onion_message_node(node_id: bytes, node_info: Optional['NodeInfo']):
+    if not node_info:
+        return False
+    return LnFeatures(node_info.features).supports(LnFeatures.OPTION_ONION_MESSAGE_OPT)
+
+
+def encrypt_onionmsg_tlv_hops_data(hops_data, hop_shared_secrets):
+    """encrypt unencrypted onionmsg_tlv.encrypted_recipient_data for hops with blind_fields"""
+    num_hops = len(hops_data)
+    for i in range(num_hops):
+        if hops_data[i].tlv_stream_name == 'onionmsg_tlv' and 'encrypted_recipient_data' not in hops_data[i].payload:
+            # construct encrypted_recipient_data from blind_fields
+            encrypted_recipient_data = encrypt_encrypted_data_tlv(shared_secret=hop_shared_secrets[i], **hops_data[i].blind_fields)
+            hops_data[i].payload['encrypted_recipient_data'] = {'encrypted_recipient_data': encrypted_recipient_data}
+
+
+async def create_onion_message_route_to(lnwallet: 'LNWallet', node_id: bytes) -> List[PathEdge]:
+    """Constructs a route to the destination node_id, first by starting with peers with existing channels,
+       and if no route found, opening a direct peer connection if node_id is found with an address in
+       channel_db."""
+    # TODO: is this the proper way to set up my_sending_channels?
+    my_active_channels = [
+        chan for chan in lnwallet.channels.values() if
+        chan.is_active() and not chan.is_frozen_for_sending()]
+    my_sending_channels = {chan.short_channel_id: chan for chan in my_active_channels
+                           if chan.short_channel_id is not None}
+    # strat1: find route to introduction point over existing channel mesh
+    # NOTE: nodes that are in channel_db but are offline are not removed from the set
+    if lnwallet.network.path_finder:
+        if path := lnwallet.network.path_finder.find_path_for_payment(
+            nodeA=lnwallet.node_keypair.pubkey,
+            nodeB=node_id,
+            invoice_amount_msat=10000,  # TODO: do this without amount constraints
+            node_filter=lambda x, y: True if x == lnwallet.node_keypair.pubkey else is_onion_message_node(x, y),
+            my_sending_channels=my_sending_channels
+        ): return path
+
+    # strat2: dest node has host:port in channel_db? then open direct peer connection
+    if lnwallet.channel_db:
+        if peer_addr := lnwallet.channel_db.get_last_good_address(node_id):
+            peer = await lnwallet.add_peer(str(peer_addr))
+            await peer.initialized
+            return [PathEdge(short_channel_id=None, start_node=None, end_node=node_id)]
```

Comment:
Is this really needed?  Bolt-04 suggests that onion messages should use already existing connections, not create new ones:

> Onion messages allow peers to use existing connections 

I would rather not have methods that create extra peer connections as side-effect, unless there is a clear use case.